### DISCUSSION TITLE
Custom tex for book

### DISCUSSION THREAD

**Discussion by nicole-brewer** [quarto-dev/quarto-cli#6513] [Q&A]
### Description

Hi! I'll start by saying that Quarto is amazing and I hope to use it to format my dissertation. I was originally interested in creating a formal template others could use, but then I realized that books aren't supported template formats, just pdfs. So I abandoned the idea of creating a template and switched to using a book. 

Here is my problem: because I have to follow a strict format, using the pdf option `include-in-header` (shown below) is not enough customization for me. 

```yaml
format:
  pdf: 
    include-in-header: preamble.tex
```
I need to be able to supply all my own Latex to conform to university requirements. Of course, I was successfully able to do this when I was making my own pdf template, but I've been unable to figure out if this is possible for the book format. I suspect that it isn't possible and that I should actually be posting this in feature requests.

It's nice to see, based on #4546, that I'm not the only one interested in this! Its really encouraging to see others interested in reproducible research (in the Jon Claerbout sense). Unfortunately for me, I need full control over the tex file, and I imagine that as quarto-based dissertations become more popular that I will not be the only one with strict formatting requirements that can only be met by a custom tex file. 

Thanks a lot!

**Comment by mcanouil** [comment]
Book is the type of a Quarto project.
PDF/HTML is the format used.
You can combined the two as it was done by others:
- https://github.com/fredguth/tufte-quarto

See other extensions:
- https://github.com/numbats/monash-quarto-thesis/tree/main

Project type extension is described in <https://quarto.org/docs/extensions/project-types.html>.

If you don't want to create an extension and tweak a lot of the template, then you need to look into partials (also useful when creating an extension), see <https://quarto.org/docs/journals/templates.html>.

**Comment by nicole-brewer** [comment]
This response was incredibly helpful, thank you! I now have a book project using my own pdf format that makes extensive use of pandoc templating!  I also found some really useful videos explaining lua filters and how to use them to manipulate yaml metadata. Yet there's one thing I still don't understand...

I don't understand how to get access to the `$body$` (that comes from my markdown chapters) without also importing  `$pandoc.tex()$`. I think I read that this partial is required, but it adds a lot of stuff to body that alters the original. Things like  `\addcontentsline{toc}{chapter}{Introduction}` are clearly desirable and/or inconsequential, however there is some code in there that is also causing (numbered) chapter titles that used to render as 

![Screen Shot 2023-08-15 at 9 11 33 PM](https://github.com/quarto-dev/quarto-cli/assets/20686935/c15afcc1-b206-4ccd-8513-5987131e85fe)

to have new format,

![Screen Shot 2023-08-15 at 9 11 37 PM](https://github.com/quarto-dev/quarto-cli/assets/20686935/6b12dabc-977c-47bf-98be-2f031fb4467c)

So I suppose my first question is, is there any way to avoid partials all together and still get access to all the content in my markdown chapters? and if not, how might you approach this problem? I didn't write the latex myself so re-writing it to be compatible with the quarto partials seems like a very daunting task.

Thank you so much!

**Reply by rgaiacs** [reply]
Dear @nicole-brewer,

without a full minimal example to follow, my debug steps to the numbered chapter would be

1. Check that the Quarto file is using **not** numbered chapter
2. Enable `keep-tex` in the Quarto project and inspect the generated `.tex`
   1. Is the LaTeX code `\chapter*{}`?

**Reply by nicole-brewer** [reply]
It funny what ends up being helpful when someone responds. The chapter **is numbered**, which I was using because the formatting is much closer to the original. When I use the unnumbered class on the chapter title (`# Summary {.unnumbered}`) the formatting looks much more like the Quarto default, which is *further* from what I want.

![Screen Shot 2023-08-15 at 9 11 19 PM](https://github.com/quarto-dev/quarto-cli/assets/20686935/e8a21d64-e856-42be-97a4-783981ba781b)

I have actually been using `keep-tex` which is how I know the chapter is given to me as

```tex
\bookmarksetup{startatroot}

\hypertarget{summary}{%
\chapter{Summary}\label{summary}}

In summary, this book has no content whatsoever.
```

<details>
<summary>Quarto-thesis.tex</summary>
<code>
\documentclass[12pt,letterpaper]{report}
\usepackage{natbib}
\usepackage{geometry}
\usepackage{fancyhdr}
\usepackage{afterpage}
\usepackage{graphicx}
\usepackage{amsmath,amssymb,amsbsy}
\usepackage{dcolumn,array}
\usepackage{tocloft}
\usepackage{asudis}



\providecommand{\tightlist}{%
  \setlength{\itemsep}{0pt}\setlength{\parskip}{0pt}}\usepackage{longtable,booktabs,array}
\usepackage{calc} % for calculating minipage widths
% Correct order of tables after \paragraph or \subparagraph
\usepackage{etoolbox}
\makeatletter
\patchcmd\longtable{\par}{\if@noskipsec\mbox{}\fi\par}{}{}
\makeatother
% Allow footnotes in longtable head/foot
\IfFileExists{footnotehyper.sty}{\usepackage{footnotehyper}}{\usepackage{footnote}}
\makesavenoteenv{longtable}
\usepackage{graphicx}
\makeatletter
\def\maxwidth{\ifdim\Gin@nat@width>\linewidth\linewidth\else\Gin@nat@width\fi}
\def\maxheight{\ifdim\Gin@nat@height>\textheight\textheight\else\Gin@nat@height\fi}
\makeatother
% Scale images if necessary, so that they will not overflow the page
% margins by default, and it is still possible to overwrite the defaults
% using explicit options in \includegraphics[width, height, ...]{}
\setkeys{Gin}{width=\maxwidth,height=\maxheight,keepaspectratio}
% Set default figure placement to htbp
\makeatletter
\def\fps@figure{htbp}
\makeatother
\newlength{\cslhangindent}
\setlength{\cslhangindent}{1.5em}
\newlength{\csllabelwidth}
\setlength{\csllabelwidth}{3em}
\newlength{\cslentryspacingunit} % times entry-spacing
\setlength{\cslentryspacingunit}{\parskip}
\newenvironment{CSLReferences}[2] % #1 hanging-ident, #2 entry spacing
 {% don't indent paragraphs
  \setlength{\parindent}{0pt}
  % turn on hanging indent if param 1 is 1
  \ifodd #1
  \let\oldpar\par
  \def\par{\hangindent=\cslhangindent\oldpar}
  \fi
  % set entry spacing
  \setlength{\parskip}{#2\cslentryspacingunit}
 }%
 {}
\usepackage{calc}
\newcommand{\CSLBlock}[1]{#1\hfill\break}
\newcommand{\CSLLeftMargin}[1]{\parbox[t]{\csllabelwidth}{#1}}
\newcommand{\CSLRightInline}[1]{\parbox[t]{\linewidth - \csllabelwidth}{#1}\break}
\newcommand{\CSLIndent}[1]{\hspace{\cslhangindent}#1}

\makeatletter
\makeatother
\makeatletter
\@ifpackageloaded{bookmark}{}{\usepackage{bookmark}}
\makeatother
\makeatletter
\@ifpackageloaded{caption}{}{\usepackage{caption}}
\AtBeginDocument{%
\ifdefined\contentsname
  \renewcommand*\contentsname{Table of contents}
\else
  \newcommand\contentsname{Table of contents}
\fi
\ifdefined\listfigurename
  \renewcommand*\listfigurename{List of Figures}
\else
  \newcommand\listfigurename{List of Figures}
\fi
\ifdefined\listtablename
  \renewcommand*\listtablename{List of Tables}
\else
  \newcommand\listtablename{List of Tables}
\fi
\ifdefined\figurename
  \renewcommand*\figurename{Figure}
\else
  \newcommand\figurename{Figure}
\fi
\ifdefined\tablename
  \renewcommand*\tablename{Table}
\else
  \newcommand\tablename{Table}
\fi
}
\@ifpackageloaded{float}{}{\usepackage{float}}
\floatstyle{ruled}
\@ifundefined{c@chapter}{\newfloat{codelisting}{h}{lop}}{\newfloat{codelisting}{h}{lop}[chapter]}
\floatname{codelisting}{Listing}
\newcommand*\listoflistings{\listof{codelisting}{List of Listings}}
\makeatother
\makeatletter
\@ifpackageloaded{caption}{}{\usepackage{caption}}
\@ifpackageloaded{subcaption}{}{\usepackage{subcaption}}
\makeatother
\makeatletter
\@ifpackageloaded{tcolorbox}{}{\usepackage[skins,breakable]{tcolorbox}}
\makeatother
\makeatletter
\@ifundefined{shadecolor}{\definecolor{shadecolor}{rgb}{.97, .97, .97}}
\makeatother
\makeatletter
\makeatother
\makeatletter
\makeatother

\begin{document}
%-----------------------front matter
\pagenumbering{roman}
%
% Title page.
%
% Redefine maketitle
\newlength{\fiveblanklines}\setlength{\fiveblanklines}{0.7 in}
\newlength{\tenblanklines}\setlength{\tenblanklines}{1.5 in}
\renewcommand\maketitle{
  \enlargethispage{0.5in} %make extra room where page number would go
  \thispagestyle{empty}
  \noindent
  \begin{minipage}[t][647.0truept][t]{\linewidth}
  \exactdoublespace
  \begin{center}
    Quarto-thesis\\
    by\\
    Nicole Brewer
  \end{center}
  \singlespace
  \vspace{\fiveblanklines}
  \begin{center}
    \singlespace
	A Doctor of Philosophy Presented in Partial Fulfillment\\
	of the Requirements for the Degree\\
	History and Philosophy of Science
  \end{center}
  \vspace{\tenblanklines}
  \begin{center}
    \singlespace
    Approved April 15, 2026 by the\\
    Graduate Supervisory Committee:\\ \ \\
        Manfred Laubichler, Chair\\
        Jane Maienschein, Chair\\
          \end{center}
   \vfill
   \begin{center}
   \doublespace
   ARIZONA STATE UNIVERSITY\\
   May 15, 2026
   \end{center} 
\end{minipage}
\clearpage
}


\maketitle
\doublespace
    \begin{abstract}
    Some abstract text about a very exciting (and reproducible)
    dissertation!
    \end{abstract}

    \dedicationpage{This dissertation is dedicated to my faithful
ergonomic mouse that kept me from getting carpal tunnel, xxx}

    \begin{acknowledgements}
    \dedicationpage{I'd like to thank all the wonderful people who
supported me through this difficult journey.}
    \end{acknowledgements}

\tableofcontents
% This puts the word "Page" right justified above everything else.
\addtocontents{toc}{~\hfill Page\par}
% Asking LaTeX for a new page here guarantees that the LOF is on a separate page
% after the TOC ends.
\newpage
% Making the LOT and LOF "parts" rather than chapters gets them indented at
% level -1 according to the chart: top of page 4 of the document at
% ftp://tug.ctan.org/pub/tex-archive/macros/latex/contrib/tocloft/tocloft.pdf
\addcontentsline{toc}{part}{LIST OF TABLES}
\renewcommand{\cftlabel}{Table}
\listoftables
% This gets the headers for the LOT right on the first page.  Subsequent pages
% are handled by the fancyhdr code in the asudis.sty file.
\addtocontents{lot}{Table~\hfill Page \par}
\newpage
\addcontentsline{toc}{part}{LIST OF FIGURES}
\addtocontents{toc}{CHAPTER \par}
\renewcommand{\cftlabel}{Figure}
% This gets the headers for the LOF right on the first page.  Subsequent pages
% are handled by the fancyhdr code in the asudis.sty file.
\addtocontents{lof}{Figure~\hfill Page \par}
\listoffigures

%-----------------------body
\doublespace
\pagenumbering{arabic}

\bookmarksetup{startatroot}

\hypertarget{introduction}{%
\chapter{Introduction}\label{introduction}}

This is a book created from markdown and executable code.

See Knuth (1984) for additional discussion of literate programming.

\bookmarksetup{startatroot}

\hypertarget{summary}{%
\chapter{Summary}\label{summary}}

In summary, this book has no content whatsoever.

\hypertarget{refs}{}
\begin{CSLReferences}{1}{0}
\leavevmode\vadjust pre{\hypertarget{ref-knuth84}{}}%
Knuth, Donald E. 1984. {``Literate Programming.''} \emph{Comput. J.} 27
(2): 97--111. \url{https://doi.org/10.1093/comjnl/27.2.97}.

\end{CSLReferences}

%-----------------------back matter
{\singlespace
% Making the references a "part" rather than a chapter gets it indented at
% level -1 according to the chart: top of page 4 of the document at
% ftp://tug.ctan.org/pub/tex-archive/macros/latex/contrib/tocloft/tocloft.pdf
\addcontentsline{toc}{part}{REFERENCES}
%\bibliographystyle{asudis}
\bibliography{references.bib}
%\renewcommand{\chaptername}{APPENDIX}
%\addtocontents{toc}{APPENDIX \par}
%\appendix
\end{document}

</code>
</details>

which is generated from... 
<details>
<summary>_extensions/asu-quarto-thesis/template.tex</summary>
<code>
\documentclass[12pt,letterpaper]{report}
\usepackage{natbib}
\usepackage{geometry}
\usepackage{fancyhdr}
\usepackage{afterpage}
\usepackage{graphicx}
\usepackage{amsmath,amssymb,amsbsy}
\usepackage{dcolumn,array}
\usepackage{tocloft}
\usepackage{asudis}

$pandoc.tex()$ % <<<<--------------------------- the only quarto partial I import

\begin{document}
%-----------------------front matter
\pagenumbering{roman}
%
% Title page.
%
% Redefine maketitle
\newlength{\fiveblanklines}\setlength{\fiveblanklines}{0.7 in}
\newlength{\tenblanklines}\setlength{\tenblanklines}{1.5 in}
\renewcommand\maketitle{
  \enlargethispage{0.5in} %make extra room where page number would go
  \thispagestyle{empty}
  \noindent
  \begin{minipage}[t][647.0truept][t]{\linewidth}
  \exactdoublespace
  \begin{center}
    $title$\\
    by\\
    $author$
  \end{center}
  \singlespace
  \vspace{\fiveblanklines}
  \begin{center}
    \singlespace
	A $degree-type$ Presented in Partial Fulfillment\\
	of the Requirements for the Degree\\
	$degree-name$
  \end{center}
  \vspace{\tenblanklines}
  \begin{center}
    \singlespace
    Approved $defense-date$ by the\\
    Graduate Supervisory Committee:\\ \ \\
    $for(chair)$
    $chair$, Chair\\
    $endfor$
    $for(committee-members)$
    $committee-members$\\
    $endfor$
  \end{center}
   \vfill
   \begin{center}
   \doublespace
   ARIZONA STATE UNIVERSITY\\
   $graduation-date$
   \end{center} 
\end{minipage}
\clearpage
}


\maketitle
\doublespace
$if(abstract)$
    \begin{abstract}
    $abstract$
    \end{abstract}
$endif$

$if(dedication)$
    \dedicationpage{$dedication$}
$endif$

$if(acknowledgements)$
    \begin{acknowledgements}
    \dedicationpage{$acknowledgements$}
    \end{acknowledgements}
$endif$

\tableofcontents
% This puts the word "Page" right justified above everything else.
\addtocontents{toc}{~\hfill Page\par}
% Asking LaTeX for a new page here guarantees that the LOF is on a separate page
% after the TOC ends.
\newpage
% Making the LOT and LOF "parts" rather than chapters gets them indented at
% level -1 according to the chart: top of page 4 of the document at
% ftp://tug.ctan.org/pub/tex-archive/macros/latex/contrib/tocloft/tocloft.pdf
\addcontentsline{toc}{part}{LIST OF TABLES}
\renewcommand{\cftlabel}{Table}
\listoftables
% This gets the headers for the LOT right on the first page.  Subsequent pages
% are handled by the fancyhdr code in the asudis.sty file.
\addtocontents{lot}{Table~\hfill Page \par}
\newpage
\addcontentsline{toc}{part}{LIST OF FIGURES}
\addtocontents{toc}{CHAPTER \par}
\renewcommand{\cftlabel}{Figure}
% This gets the headers for the LOF right on the first page.  Subsequent pages
% are handled by the fancyhdr code in the asudis.sty file.
\addtocontents{lof}{Figure~\hfill Page \par}
\listoffigures

%-----------------------body
\doublespace
\pagenumbering{arabic}

$body$

%-----------------------back matter
{\singlespace
% Making the references a "part" rather than a chapter gets it indented at
% level -1 according to the chart: top of page 4 of the document at
% ftp://tug.ctan.org/pub/tex-archive/macros/latex/contrib/tocloft/tocloft.pdf
\addcontentsline{toc}{part}{REFERENCES}
%\bibliographystyle{asudis}
\bibliography{$bibliography$}
%\renewcommand{\chaptername}{APPENDIX}
%\addtocontents{toc}{APPENDIX \par}
%\appendix
\end{document}

</code>
</details>

My trouble is that I don't see anything that was included in the quarto partial that seems responsible for all this formatting (I'm no latex wizard). I there no way to avoid importing the quarto partial at all (yet still having the body with the chapters?

Even if I knew how to redefine new commands to overwrite the formatting in `$pandoc.tex()$`, I would have issues, because most of the code in my project is in a .sty file that I don't think can be migrated to the header of dis.tex (I tried that already). Meaning that I wouldn't be able to have the code in asudis.sty define commands *after* they are defined in the quarto partial pandoc.tex.

**Reply by rgaiacs** [reply]
Dear @nicole-brewer,

Thanks for the clarification.

The LaTeX snippet that you highlighted is generated by Pandoc. For example, taking

```
# Preface {.unnumbered}

This is a Quarto book.
```

as input and running `pandoc --from markdown --to latex index.qmd` the output is

```
\hypertarget{preface}{%
\section*{Preface}\label{preface}}
\addcontentsline{toc}{section}{Preface}

This is a Quarto book.

To learn more about Quarto books visit
\url{https://quarto.org/docs/books}.
```

`\section{}` and `\chapter{}` matching to `H1` is controlled by Pandoc. Pandoc also wrap heading in `\hypertarget{}` because, I presume, is the only way to provide links reliable, see https://github.com/jgm/pandoc/blob/main/src/Text/Pandoc/Writers/LaTeX.hs#L720-L747.

As far as I know, Pandoc does **not** offer the option to translate

```
# Preface {.unnumbered}
```

into

```
\section*{Preface}
```

Pandoc and Quarto offers `--include-in-header`, `--include-before-body`, and `--include-after-body`. In a **oversimplify** example, the generated LaTeX is

```
\documentclass[]{article}

% Include in header goes here

\begin{document}

% Include before body goes here

% All your .qmd files goes here

% Include after body goes here

\end{document}
```

Can you use `--include-in-header` and `--include-before-body` to achieve the output that you want?

**Reply by baptiste** [reply]
> is there any way to avoid partials all together and still get access to all the content in my markdown chapters?

Definitely: you can use a [custom template without any of the partials](https://quarto.org/docs/journals/templates.html#templates). You'll probably find that you need to bring back some of the macros defined by pandoc or quarto, but you can include them as needed in your own tex template file. 

For what it's worth, I [recently had troubles with those `\hypertarget` things](https://github.com/quarto-dev/quarto-cli/discussions/5696#discussioncomment-6018305) when using a book publisher template, so I ended up defining dummy macros for it:

```
\newcommand\hypertarget[2]{#2}
\newcommand\texorpdfstring[2]{#1}
\newcommand\bookmarksetup[1]{}
\newcommand\href[1]{#1}
```
